program/src/state.rs

```diff
@@ -65,24 +78,208 @@ pub trait SlashingProofData<'a> {
+
+    /// Pack the proof data into a raw data buffer
+    fn pack_proof(self) -> Vec<u8>;
+
     /// Zero copy from raw data buffers and initialize any context
-    fn unpack<'b>(
+    fn unpack_proof_and_context<'b>(
         proof_account_data: &'a [u8],
         instruction_data: &'a [u8],
-        account_info_iter: &'a mut Iter<'_, AccountInfo<'b>>,
+        accounts: &SlashingAccounts<'a, 'b>,
     ) -> Result<(Self, Self::Context), SlashingError>
     where
         Self: Sized;
 
     /// Verification logic for this type of proof data
     fn verify_proof(
-        self,
+        &self,
         context: Self::Context,
         slot: Slot,
         pubkey: &Pubkey,
     ) -> Result<(), SlashingError>;
 }
 
+/// Accounts relevant for the slashing program
+pub struct SlashingAccounts<'a, 'b> {
+    pub(crate) proof_account: &'a AccountInfo<'b>,
+    pub(crate) violation_pda_account: &'a AccountInfo<'b>,
+    pub(crate) instructions_sysvar: &'a AccountInfo<'b>,
+    pub(crate) system_program_account: &'a AccountInfo<'b>,
+}
+
+impl<'a, 'b> SlashingAccounts<'a, 'b> {
+    pub(crate) fn new<I>(account_info_iter: &mut I) -> Result<Self, ProgramError>
+    where
+        I: Iterator<Item = &'a AccountInfo<'b>>,
+    {
+        let res = Self {
+            proof_account: next_account_info(account_info_iter)?,
+            violation_pda_account: next_account_info(account_info_iter)?,
+            instructions_sysvar: next_account_info(account_info_iter)?,
+            system_program_account: next_account_info(account_info_iter)?,
+        };
+        if !sysvar::instructions::check_id(res.instructions_sysvar.key) {
+            return Err(ProgramError::from(SlashingError::MissingInstructionsSysvar));
+        }
+        if !system_program::check_id(res.system_program_account.key) {
+            return Err(ProgramError::from(
+                SlashingError::MissingSystemProgramAccount,
+            ));
+        }
+        Ok(res)
+    }
+
+    pub(crate) fn proof_account(&self) -> &Pubkey {
+        self.proof_account.key
+    }
+
+    fn violation_account(&self) -> &Pubkey {
+        self.violation_pda_account.key
+    }
+
+    fn violation_account_exists(&self) -> Result<bool, ProgramError> {
+        Ok(!self.violation_pda_account.data_is_empty()
+            && check_id(self.violation_pda_account.owner)
+            && ViolationReport::version(&self.violation_pda_account.try_borrow_data()?) > 0)
+    }
+
+    fn write_violation_report<T>(
+        &self,
+        report: ViolationReport,
+        proof: T,
+    ) -> Result<(), ProgramError>
+    where
+        T: SlashingProofData<'a>,
+    {
+        self.violation_pda_account.try_borrow_mut_data()?
+            [0..std::mem::size_of::<ViolationReport>()]
+            .copy_from_slice(bytemuck::bytes_of(&report));
+        self.violation_pda_account.try_borrow_mut_data()?[std::mem::size_of::<ViolationReport>()..]
+            .copy_from_slice(&T::pack_proof(proof));
```

Comment:
Sorry, I should have mentioned this earlier, but if the account doesn't have enough space, this will panic -- wanna add a length check beforehand to avoid potentially panicking?